### PR TITLE
feat(REST): RHICOMPL-3904 generate APIv2 spec

### PR DESCRIPTION
**Pull request by romanblanco**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: spec/api/v2/schemas/metadata.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            sort_by: {
+              type: 'string',
+              examples: %w[asc desc],
+              default: 'version:asc'
+            },
+            # FIXME: compatibility with the V1 logic from the common concerns
+            relationships: {
+              type: 'boolean',
+              examples: [false, true]
+            },
```
Comment (skateman) [inline spec/api/v2/schemas/metadata.rb:37]: Nope, you can drop this

File: spec/api/v2/schemas/metadata.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            offset: {
+              type: 'number',
+              minimum: 1,
+              default: 1,
+              examples: [1, 3]
+            },
+            sort_by: {
+              type: 'string',
+              examples: %w[asc desc],
```
Comment (skateman) [inline spec/api/v2/schemas/metadata.rb:30]: ??

File: spec/api/v2/schemas/metadata.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            relationships: {
+              type: 'boolean',
+              examples: [false, true]
+            },
+            filter: {
+              type: 'string',
+              examples: ["name='Standard System Security Profile for Fedora'"]
+            }
+          },
+          required: %w[limit offset]
```
Comment (skateman) [inline spec/api/v2/schemas/metadata.rb:43]: It's not really required, we have default values...

File: spec/api/v2/schemas/metadata.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+              type: 'string'
+            },
+            previous: {
+              type: 'string'
+            },
+            next: {
+              type: 'string'
+            }
+          },
+          required: %w[first last]
```
Comment (skateman) [inline spec/api/v2/schemas/metadata.rb:62]: Why is this required? We return these as a response...

File: spec/swagger_helper.rb [hunk 5]
```diff
@@ -4,12 +4,15 @@
 require 'webmock/rspec'
 require './spec/api/v1/openapi'
 require './spec/api/v1/schemas/util'
+require './spec/api/v2/openapi'
 
 include Api::V1::Schemas::Util # rubocop:disable Style/MixinUsage
+include Api::V2::Schemas::Util # rubocop:disable Style/MixinUsage
```
Comment (skateman) [inline spec/swagger_helper.rb:10]: Please use the multiline disable instead.

**Review by skateman — CHANGES_REQUESTED** [review]

File: spec/api/v2/schemas/metadata.rb (hunk 1 above)
Comment (romanblanco) [reply spec/api/v2/schemas/metadata.rb:37]: The new endpoint still generates `relationships`.

Comment (skateman) [reply spec/api/v2/schemas/metadata.rb:37]: How and why? The parameter handling doesn't accept this...

Comment (skateman) [reply spec/api/v2/schemas/metadata.rb:36]: ^^ constant

File: spec/api/v2/schemas/metadata.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+              examples: ["name='Standard System Security Profile for Fedora'"]
+            }
+          }
+        }.freeze
+
+        LINKS = {
+          type: 'object',
+          properties: {
+            first: {
+              type: 'string'
```
Comment (skateman) [inline spec/api/v2/schemas/metadata.rb:48]: Isn't there an URL type in rswag?

File: spec/api/v2/schemas/security_guides.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+module Api
+  module V2
+    module Schemas
+      module SecurityGuides
+        extend Api::V2::Schemas::Util
+
+        SECURITY_GUIDE = {
+          type: 'object',
+          required: %w[ref_id title version],
```
Comment (skateman) [inline spec/api/v2/schemas/security_guides.rb:13]: Why is the `os_major_version` not required?

File: spec/integration/security_guides_spec.rb [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+      description 'Lists all Security guides'
+      operationId 'ListSecurityGuides'
+
+      parameter name: :search, in: :query, required: false,
+                description: 'Query string compliant with '\
+                '<a href="https://github.com/wvanbergen/scoped_search/wiki/Query-language" '\
+                'target="_blank" title="github.com/wvanbergen/scoped_search">scoped_search query language</a>.'\
+                '<br>'\
+                'Security guides are searchable using attributes '\
+                '`ref_id`, `title`, `version` and `os_major_version`.'\
```
Comment (skateman) [inline spec/integration/security_guides_spec.rb:53]: Shouldn't this be generated?

File: app/models/v2/security_guide.rb [hunk 9]
```diff
@@ -45,7 +45,7 @@ class SecurityGuide < ApplicationRecord
     sortable_by :os_major_version, SORT_BY_OS_MAJOR_VERSION
 
     def os_major_version
-      ref_id[/(?<=RHEL-)\d/].to_i
+      ref_id[/(?<=RHEL-)\d+/].to_i
```
Comment (skateman) [inline app/models/v2/security_guide.rb:48]: Rebase? 🤔

File: public/api/compliance/v2/openapi.json [hunk 10]
```diff
@@ -0,0 +1 @@
+../../../../swagger/v2/openapi.json
```
Comment (skateman) [inline public/api/compliance/v2/openapi.json:1]: We should not expose this yet

File: spec/api/v2/schemas/metadata.rb [hunk 11]
```diff
@@ -33,7 +33,8 @@ module Metadata
             relationships: {
               type: :boolean,
               enum: [false],
-              examples: [false]
+              examples: [false],
+              description: 'nakokot'
```
Comment (skateman) [inline spec/api/v2/schemas/metadata.rb:37]: 🤣

**Review by skateman — CHANGES_REQUESTED** [review]

File: spec/api/v2/schemas/metadata.rb (hunk 11 above)
Comment (petrblaho) [reply spec/api/v2/schemas/metadata.rb:37]: :-)

Comment (RoamingNoMaD) [reply spec/api/v2/schemas/metadata.rb:37]: 😆

File: spec/api/v2/schemas.rb [hunk 12]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+require './spec/api/v2/schemas/security_guides'
+require './spec/api/v2/schemas/metadata'
+
+module Api
+  module V2
+    module Schemas
+      include Metadata
+      include SecurityGuides
```
Comment (skateman) [inline spec/api/v2/schemas.rb:10]: This feels like could be autoloaded in a similar fashion using `descendants` or some other class magic.

File: spec/api/v2/schemas/metadata.rb [hunk 13]
```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Api
+  module V2
+    module Schemas
+      module Metadata
+        UUID = { type: :string, format: :uuid }.freeze
+        METADATA = {
```
Comment (skateman) [inline spec/api/v2/schemas/metadata.rb:8]: Logical separation, if you're doing it below, do it here as well:
```suggestion
        UUID = { type: :string, format: :uuid }.freeze
        
        METADATA = {
```

File: spec/api/v2/schemas/security_guides.rb [hunk 14]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+require './spec/api/v2/schemas/util'
```
Comment (skateman) [inline spec/api/v2/schemas/security_guides.rb:3]: If you do the recursive require in the correct order above, you might not need this.

File: spec/integration/security_guides_spec.rb [hunk 15]
```diff
@@ -0,0 +1,130 @@
+# frozen_string_literal: true
+
+require 'swagger_helper'
+
+describe 'Security Guides', swagger_doc: 'v2/openapi.json' do
+  before do
+    allow_any_instance_of(Authentication).to receive(:identity_header).and_return(
```
Comment (skateman) [inline spec/integration/security_guides_spec.rb:7]: Please stop using `_any_instance_of` methods, first of all it is an antipattern and second it is slowing down testing. There are plenty of [mocks](https://github.com/RedHatInsights/compliance-backend/blob/master/spec/controllers/v2/application_controller_spec.rb#L13-L14) already in existing tests that you can inspire from.

File: spec/integration/security_guides_spec.rb [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+              {
+                'insights': {
+                  'is_entitled': true
+                }
+              }
+          }.with_indifferent_access.to_json
+        )
+      )
+    )
+    stub_rbac_permissions(Rbac::COMPLIANCE_ADMIN, Rbac::INVENTORY_HOSTS_READ)
```
Comment (skateman) [inline spec/integration/security_guides_spec.rb:36]: I'd logically separate this line:
```suggestion
    )
    
    stub_rbac_permissions(Rbac::COMPLIANCE_ADMIN, Rbac::INVENTORY_HOSTS_READ)
```

File: spec/integration/security_guides_spec.rb [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+  end
+
+  path '/security_guides' do
+    get 'List all Security Guides' do
+      before do
+        [7, 8, 9].each do |os_version|
```
Comment (skateman) [inline spec/integration/security_guides_spec.rb:42]: I think we should be pulling these numbers from the `SupportedSsg` "model" so we don't have to change anything here it if a new version of RHEL comes out.

File: spec/integration/security_guides_spec.rb [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+      parameter name: :search,
+                in: :query,
+                required: false,
+                description: 'Query string compliant with ' \
+                  '<a href="https://github.com/wvanbergen/scoped_search/wiki/Query-language" ' \
+                  'target="_blank" title="github.com/wvanbergen/scoped_search">scoped_search query language</a>.' \
+                  '<br>' \
+                  'Security guides are searchable using attributes ' \
+                  "#{V2::SecurityGuide.scoped_search.fields.keys.map { |k| "`#{k}`" }.to_sentence}" \
+                  '<br><br>(e.g.: `(version=0.1.47 AND os_major_verision < 8)`)',
```
Comment (skateman) [inline spec/integration/security_guides_spec.rb:79]: 1. this is ugly
2. should be a parametrized method as it will be used elsewhere

File: spec/integration/security_guides_spec.rb [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+      parameter name: :sort_by,
+                in: :query,
+                required: false,
+                description: 'An array of fields with an optional direction ' \
+                  '(`<key>:asc` or `<key>:desc`) to sort the results.<br>' \
+                  'If no direction is selected, `<key>:asc` is used by default.',
+                schema: {
+                  type: :array,
+                  items: { enum: sort_combinations(V2::SecurityGuide) }
+                }
```
Comment (skateman) [inline spec/integration/security_guides_spec.rb:69]: This should be probably a parametrized method as it will be used across endpoints.

File: spec/swagger_helper.rb [hunk 20]
```diff
@@ -4,12 +4,17 @@
 require 'webmock/rspec'
 require './spec/api/v1/openapi'
 require './spec/api/v1/schemas/util'
+require './spec/api/v2/openapi'
 
-include Api::V1::Schemas::Util # rubocop:disable Style/MixinUsage
+# rubocop:disable Style/MixinUsage
+include Api::V1::Schemas::Util
+include Api::V2::Schemas::Util
```
Comment (skateman) [inline spec/swagger_helper.rb:11]: Rubocop's complaints are real here, what methods are needed from these modules in the current context and why can't they be called statically?

File: swagger/v2/openapi.json [hunk 21]
```diff
@@ -0,0 +1,350 @@
...
+            "minimum": 7,
+            "examples": [
+              7
+            ]
+          }
+        }
+      }
+    }
+  }
+}
```
Comment (skateman) [inline swagger/v2/openapi.json:350]: Can we do something about the missing EOL?

**Review by skateman — CHANGES_REQUESTED** [review]

File: spec/api/v2/schemas/types.rb [hunk 22]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+require './spec/api/v1/schemas/util'
```
Comment (skateman) [inline spec/api/v2/schemas/types.rb:3]: 🚒 🚒 🚒 🚒 🚒 🚒 🚒 🚒 🚒 🚒 🚒 🚒 🚒 🚒 🚒 🚒

File: spec/integration/security_guides_spec.rb [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+      valid?: true,
+      org_id: '1234',
+      identity: { org_id: '1234' },
+      raw: nil
+    )
+  end
+
+  before do
+    allow(Insights::Api::Common::IdentityHeader).to receive(:new).and_return(identity_header)
+
```
Comment (skateman) [inline spec/integration/security_guides_spec.rb:18]: No need for logical separation here

File: spec/swagger_helper.rb [hunk 24]
```diff
@@ -4,12 +4,14 @@
 require 'webmock/rspec'
 require './spec/api/v1/openapi'
 require './spec/api/v1/schemas/util'
+require './spec/api/v2/openapi'
```
Comment (skateman) [inline spec/swagger_helper.rb:7]: Maybe like this?
```suggestion
# require './spec/api/v2/schemas/util'
require './spec/api/v2/openapi'
```

**Review by skateman — CHANGES_REQUESTED** [review]

**Review by xmicha82 — APPROVED** [review]
LGTM

File: spec/api/v2/schemas/metadata.rb [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+              examples: %w[version:asc]
+            },
+            # FIXME: compatibility with the V1 logic from the common concerns
+            relationships: {
+              type: :boolean,
+              enum: [false],
+              examples: [false],
+              description: 'The attribute represents a return value, that can be either false or unset'
+            },
+            filter: {
```
Comment (romanblanco) [inline spec/api/v2/schemas/metadata.rb:40]: @xmicha82 are we expecting `filter` to be returned by our metadata?

File: spec/api/v2/schemas/metadata.rb [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+              default: 10,
+              examples: [10, 100],
+              description: 'Maximum number of items to return'
+            },
+            offset: {
+              type: :number,
+              minimum: 0,
+              default: 0,
+              examples: [15, 90],
+              description: 'Offset of the first item returned'
```
Comment (skateman) [inline spec/api/v2/schemas/metadata.rb:31]: https://github.com/RedHatInsights/compliance-backend/pull/1765/files#diff-1e8cf3de001b876faaf2655349c90959c6a725c8ead4744a692aa285e986a4ffR110

inconsistency...please check the rest of the params as well

Comment (romanblanco) [reply spec/api/v2/schemas/metadata.rb]: See [latest changes](https://github.com/RedHatInsights/compliance-backend/compare/ee4ccd88988d1e477561e6cee4ce92d9e5cb3780..a533924bdcf94b4de1291330dd855f69d5e93f09)

**Review by skateman — APPROVED** [review]
The Seal of Approval
